src/Build/Utilities/NuGetFrameworkWrapper.cs

```diff
@@ -133,5 +156,106 @@ public string FilterTargetFrameworks(string incoming, string filter)
                 });
             }
         }
+
+#if FEATURE_APPDOMAIN
+        /// <summary>
+        /// A null-returning InitializeLifetimeService to give the proxy an infinite lease time.
+        /// </summary>
+        public override object InitializeLifetimeService() => null;
+
+        private static AppDomainSetup CreateAppDomainSetup(string assemblyDirectory)
+        {
+            string assemblyPath = Path.Combine(assemblyDirectory, NuGetFrameworksFileName);
+            AssemblyName assemblyName;
+            try
+            {
+                assemblyName = AssemblyName.GetAssemblyName(assemblyPath);
+            }
+            catch
+            {
+                // Return null to fall back to loading into the default AppDomain using LoadFile.
+                return null;
+            }
+
+            byte[] publicKeyToken = assemblyName.GetPublicKeyToken();
+            StringBuilder publicKeyTokenString = new(publicKeyToken.Length * 2);
+            for (int i = 0; i < publicKeyToken.Length; i++)
+            {
+                publicKeyTokenString.Append(publicKeyToken[i].ToString("x2", CultureInfo.InvariantCulture));
+            }
+
+            // Create an app.config for the AppDomain. We expect the AD to host the currently executing assembly Microsoft.Build,
+            // NuGet.Frameworks, and Framework assemblies. It is important to use the same binding redirects that were used when
+            // NGENing MSBuild for the native images to be used.
+            string configuration = $@"<?xml version=""1.0"" encoding=""utf-8""?>
```

Comment:
This could alternatively be done by reading the configuration of the current AD from MSBuild.exe.config, inserting the NuGet.Frameworks-related entries, and converting the whole thing to bytes to be passed to `AppDomainSetup.SetConfigurationBytes`. It would be more robust at expense of performance as it would be manipulating a much larger config and doing I/O to read the config file off disk. I'm curious what reviewers think.

Reply:
Can we use `"""` verbatim string here to get rid of double quotation?

Otherwise - I'm not superexcited having this baked into the code, but looking on the MSBuild config, this section changes rarely. So the performance tradeoff is acceptable from my point of view.

I'm just thinking if this code and MSBuild/app.config should have comments cross-mentioning each other (in case there would be some need for change)

Reply:
Great point about `"""`! We could use a source generator to build the string from app.config at build time. If we don't eliminate it completely per [Rainer's comment](https://github.com/dotnet/msbuild/pull/9446#discussion_r1408352134).

Reply:
I wouldn't use a Roslyn source generator for this, but a script in a prebuild step could be pretty nice . . . if it ever changes and causes problems.

Reply:
Can you elaborate on the reasons why you suggest a pre-build step instead of a source generator? Curious if it's because we currently have no generators so the barrier to introduce the first one wouldn't be justified, or if it's not the right tool for this use case.

Reply:
Three main reasons:

1. It's just not necessary--the inputs to the generation don't include any source code.
2. incrementality: while there are incremental source generators, getting to that point still means you have to run the compiler. Do it in a separate target and let the compiler not be invoked in more situations
3. It's still not a great experience to have a source generator built in the same repo as its use.

Reply:
I've added the functionality as a custom target calling `XmlPeek` to extract the binding redirect from app.config. And it uses raw string literals (`"""`) too!